### PR TITLE
Typescript refactor

### PR DESCRIPTION
**Pull request by antonyfaris**
I noticed that all other Astro projects are in Typescript that's why I did this refactor.
I commented out some code instead of removing it, to keep it as a backup because the .ts files are technically new files and is more convenient. I also tried to add as many type guards as possible to get better Intellisense. I will need review on this.
Other things to keep in mind:

### Testing
`yarn build`
`yarn test`
or in watch mode:
`yarn dev`
`yarn test:w`

### Debugging
[From ava:](https://github.com/avajs/ava/blob/main/docs/recipes/debugging-with-vscode.md#debugging-with-the-debug-terminal)
You can use VS Code's “JavaScript Debug Terminal” to automatically debug AVA run on the command-line.

1. From the Command Palette (<kbd>F1</kbd> or <kbd>command + shift + p</kbd> / <kbd>control + shift + p</kbd>), run `Debug: JavaScript Debug Terminal`
2. Run `npx ava` in the terminal

### Publishing
Now is required to build first. I tried to make the necessary changes to package.json and the github workflows but I **definitely** gonna need  review.

### CODE REVIEW

**Comment by antonyfaris** [comment]
> lgtm, let me know when the conflict is resolve and I'll review.

Done!

File: package.json [hunk 1]
```diff
@@ -40,11 +42,18 @@
   },
   "devDependencies": {
     "@changesets/cli": "^2.16.0",
+    "@rollup/plugin-commonjs": "^21.0.0",
+    "@rollup/plugin-node-resolve": "^13.0.5",
     "@types/prettier": "^2.4.1",
-    "ava": "^4.0.0-alpha.2",
+    "ava": "4.0.0-alpha.2",
```
Comment (antonyfaris) [inline package.json:48]: I'm currently having problems with the new version. I think we can solve this in another pr.

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    /* Projects */
+    // "incremental": true,                              /* Enable incremental compilation */
+    // "composite": true,                                /* Enable constraints that allow a TypeScript project to be used with project references. */
+    // "tsBuildInfoFile": "./",                          /* Specify the folder for .tsbuildinfo incremental compilation files. */
+    // "disableSourceOfProjectReferenceRedirect": true,  /* Disable preferring source files instead of declaration files when referencing composite projects */
+    // "disableSolutionSearching": true,                 /* Opt a project out of multi-project reference checking when editing. */
+    // "disableReferencedProjectLoad": true,             /* Reduce the number of projects loaded automatically by TypeScript. */
+
+    /* Language and Environment */
+    "target": "es5" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
```
Comment (drwpow) [inline tsconfig.json:14]: es5 is pretty old! I’d be OK shipping ES2020 or ES2019 or something more reasonable. There can be a lot of weirdness trying to transpile for old targets.

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,100 @@
+{
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig.json to read more about this file */
+
+    /* Projects */
+    // "incremental": true,                              /* Enable incremental compilation */
```
Comment (drwpow) [inline tsconfig.json:6]: Not blocking, but IMO TypeScript ships with autocomplete/documentation on these so I find the commented values to be noisy. But don’t care strongly enough to request they be removed; I’ll let the author decide.

**Comment by drwpow** [comment]
Awesome work!

File: tsconfig.json (hunk 2 above)
Comment (matthewp) [reply tsconfig.json:14]: Yeah I agree with this, @antonyfaris ok with bumping this up?

**Comment by drwpow** [comment]
I’d strongly suggest adding a `build:` task to .github/workflows/ci.yml. Follow the format of the other two, except it runs `yarn build` instead. That way we can be confident that every PR builds successfully and without errors.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -20,6 +20,7 @@ jobs:
       - run: yarn --frozen-lockfile --ignore-engines
         env:
           CI: true
+      - run: yarn build
```
Comment (drwpow) [inline .github/workflows/ci.yml:23]: 👍🏻

File: package.json [hunk 5]
```diff
@@ -2,11 +2,10 @@
   "name": "prettier-plugin-astro",
   "version": "0.0.9",
   "description": "A Prettier Plugin for formatting Astro files",
-  "main": "src/index.js",
+  "main": "dist/index.js",
   "files": [
-    "src/**"
+    "dist/**"
```
Comment (drwpow) [inline package.json:7]: Good catch! This is an easy one to miss

**Review by drwpow — APPROVED** [review]
This seems to work and build correctly for me!

**Review by drwpow — CHANGES_REQUESTED** [review]
Sorry—I left several scattered comments and the status wasn’t clear. I’d like to request a change **only to `target: "es5"` and nothing else.** Once we update that to `ES2019` or something similar, happy to approve & merge. Otherwise it looks great!

File: tsconfig.json (hunk 2 above)
Comment (antonyfaris) [reply tsconfig.json:14]: Yes, I will do it.

**Review by drwpow — APPROVED** [review]
Amazing work! Thanks for making that target change.

**Comment by antonyfaris** [comment]
> Sorry—I left several scattered comments and the status wasn’t clear. I’d like to request a change **only to `target: "es5"` and nothing else.** Once we update that to `ES2019` or something similar, happy to approve & merge. Otherwise it looks great!

Done. I think there is still room for improvement in this ts setup. I will keep working on this.

**Comment by drwpow** [comment]
Cool. Yes please feel free to make any other changes you’d like in followup PRs! Just wanted to get this merged since this was a lot of work!

**Pull request merged**